src/rounding.rs

```diff
@@ -0,0 +1,297 @@
+use crate::{
+    options::{TemporalRoundingMode, TemporalUnsignedRoundingMode},
+    TemporalError, TemporalResult,
+};
+
+use std::{
+    cmp::Ordering,
+    ops::{Div, Neg},
+};
+
+use num_traits::{ConstZero, Euclid, FromPrimitive, Signed};
+
+trait Roundable: PartialOrd {
+    fn is_exact(&self) -> bool;
+    fn cmp_rem(&self) -> Option<Ordering>;
+    fn is_cardinal(&self) -> bool;
+    fn result_floor(&self) -> u64;
+    fn result_ceil(&self) -> u64;
+}
+
+pub(crate) trait Round {
+    fn round(&self, mode: TemporalRoundingMode) -> i128;
+    fn round_as_positive(&self, mode: TemporalRoundingMode) -> u64;
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+pub(crate) struct IncrementRounder<
+    T: Euclid + Div + PartialOrd + Signed + FromPrimitive + ConstZero + Copy,
+> {
+    sign: bool,
+    dividend: T,
+    divisor: T,
+}
+
+impl<T: Euclid + PartialOrd + Signed + FromPrimitive + ConstZero + Copy> IncrementRounder<T> {
+    // ==== PUBLIC ====
+
+    pub(crate) fn from_potentially_negative_parts(number: T, increment: T) -> Self {
+        Self {
+            sign: number / increment > T::ZERO,
+            dividend: number,
+            divisor: increment,
+        }
+    }
+
+    pub(crate) fn from_positive_parts(number: T, increment: T) -> Self {
+        debug_assert!(number / increment > T::ZERO);
+
+        Self {
+            sign: true,
+            dividend: number,
+            divisor: increment,
+        }
+    }
+
+    // ==== PRIVATE ====
+
+    #[inline]
+    fn quotient_abs(&self) -> T {
+        (self.dividend / self.divisor).abs()
+    }
+}
+
+impl Round for IncrementRounder<i128> {
+    fn round(&self, mode: TemporalRoundingMode) -> i128 {
+        let unsigned_rounding_mode = mode.get_unsigned_round_mode(self.sign);
+        let mut rounded = apply_unsigned_rounding_mode(self, unsigned_rounding_mode) as i128;
+        if !self.sign {
+            rounded = rounded.neg();
+        }
+        rounded * self.divisor
+    }
+
+    fn round_as_positive(&self, mode: TemporalRoundingMode) -> u64 {
+        let unsigned_rounding_mode = mode.get_unsigned_round_mode(self.sign);
+        let rounded = apply_unsigned_rounding_mode(self, unsigned_rounding_mode);
+        rounded * self.divisor as u64
+    }
+}
```

Comment:
I'd suggest making this an impl for `IncrementRounder<T: Roundable>` instead, and using any of the casting traits in [`num_traits`](https://docs.rs/num-traits/latest/num_traits/cast/index.html) to make it fully generic.

Reply:
:confused: the only issue is that `Roundable` can't be implemented directly on `i128` it has to be implemented on the dividend/divisor pair

Reply:
Yes, I'm just thinking that those operations don't really need an intermediate struct, they can just be methods that take the divisor by value.

`Roundable::is_exact(dividend: T, divisor: T) -> bool`